File: 1046-last-stone-weight/1046-last-stone-weight.py
import heapq
class Solution(object):
    def lastStoneWeight(self, stones):
        maxheap = [-stone for stone in stones]
        heapq.heapify(maxheap)

        while len(maxheap) > 1:
            y = -heapq.heappop(maxheap)
            x = -heapq.heappop(maxheap)

            if y != x:
                heapq.heappush(maxheap, -(y - x))

        return -maxheap[0] if maxheap else 0
       

        

File: 1046-last-stone-weight/1046-last-stone-weight.js
/**
 * @param {number[]} stones
 * @return {number}
 */
var lastStoneWeight = function (stones) {
    const heap = new MaxHeap();

    for (let s of stones) {
        heap.push(s);
    }

    while (heap.size() > 1) {
        let y = heap.pop();
        let x = heap.pop();

        if (y !== x) {
            heap.push(y - x);
        }
    }

    return heap.size() ? heap.pop() : 0;
};